torch/distributed/algorithms/model_averaging/hierarchical_model_averager.py

```diff
@@ -145,14 +145,29 @@ def _find_process_group(self):
 
     def average_parameters(self, params):
         r"""
-        Averages parameters if ``step`` is no less than ``warmup_steps``
+        Averages parameters or parameter groups of an optimizer if ``step`` is no less than ``warmup_steps``
         and it can be divided by a period in the keys of ``period_process_group_dict``,
         where ``step`` is increased by 1 at each iteration in the training loop.
         If ``step`` can be divided by multiple periods in the keys of ``period_process_group_dict``,
         only the largest period is used, and the corresponding process group is used for averaging parameters.
+        params: average model.parameters() or parameter groups of an optimizer
         """
         if self.step >= self.warmup_steps:
             group = self._find_process_group()
             if group is not None:
-                utils.average_parameters(iter(params), group)
+                filter_params = []
+                for param in params:
+                    if isinstance(param, torch.nn.Parameter):
+                        # model.parameters() input
+                        param_data = param
+                        if param_data.grad is not None:
+                            filter_params.append(param_data)
+                    elif isinstance(param, dict):
+                        # optimzer.param_groups input
+                        for param_data in param["params"]:
+                            if param_data.grad is not None:
+                                filter_params.append(param_data)
+                    else:
+                        raise NotImplementedError
+                utils.average_parameters(filter_params, group)
```

Comment:
does not have to be in this diff, but let's definitely deduplicate this code in follow up PRs.

Reply:
i supposed you mean to optimize this duplicate cod between
 HierarchicalModelAverager.average_parameters() and PeriodicModelAverager.average_parameters() in the following pr.
so i doest not change this code.